default plan admission controller: filter list of service plans/service classes by the class name

jboyd01 opened the pull request:
when getting list of plans to determine default plan name, limit the service classes and plan names we consider by only including those associated with the specified class name.

fixes #1346

pmorie commented:
needs unit coverage

jboyd01 commented:
@pmorie let's discuss.  I added additional unit test but I'm not in favor of this approach.

vaikas-google commented:
Looks like an import mismatch somewhere? Bad merge?

vaikas-google commented on plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go:
```diff
@@ -102,11 +103,13 @@ func (d *defaultServicePlan) Admit(a admission.Attributes) error {
 
-	// implement field selector
-
 	// loop over all service plans accumulate into slice
```
Maybe update this comment? get all the serviceplans for this class?

jboyd01 replied:
sure!

vaikas-google commented on plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go:
```diff
@@ -160,25 +164,51 @@ func (d *defaultServicePlan) Validate() error {
-	servicePlans, err := d.scClient.List(listOpts)
+func (d *defaultServicePlan) getClusterServiceClassByExternalName(a admission.Attributes, scName string) (*servicecatalog.ClusterServiceClass, error) {
+	glog.V(4).Infof("Fetching serviceclass filtered by class external name %q", scName)
+	fieldSet := fields.Set{
```
Just curious why replace a single line with 4?

pmorie replied:
This method escapes the field selector if necessary

jboyd01 replied:
I picked up this initial change from Paul and I had to ask him.  He says this "escapes the field selector if necessary" and its a best practice per api machinery guys

vaikas-google commented on plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go:
```diff
@@ -160,25 +164,51 @@ func (d *defaultServicePlan) Validate() error {
+// ServicePlans for the specified service class name
+func (d *defaultServicePlan) getClusterServicePlansByClusterServiceClassName(scName string) ([]servicecatalog.ClusterServicePlan, error) {
+	glog.V(4).Infof("Fetching ClusterServicePlans by class name %q", scName)
+	fieldSet := fields.Set{
```
same question here, why not just construct the one liner, I find that cleaner to read since we only have a single entry. Again, just curious on the benefit here.

jboyd01 commented:
I addressed Ville's review and remerged to address compilation issue with unit test.  All set now.

pmorie commented on plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go:
```diff
@@ -57,6 +57,7 @@ func Register(plugins *admission.Plugins) {
 	*admission.Handler
 	scClient servicecataloginternalversion.ClusterServiceClassInterface
+	spClient servicecataloginternalversion.ClusterServicePlanInterface
 	spLister internalversion.ClusterServicePlanLister
```
Do we still need or use the lister anymore?

pmorie commented on plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go:
```diff
@@ -103,11 +104,13 @@ func (d *defaultServicePlan) Admit(a admission.Attributes) error {
-	// loop over all service plans accumulate into slice
-	plans, err := d.spLister.List(labels.Everything())
-	// TODO filter `plans` down to only those owned by `sc`.
+	// loop over all service plans for the specified classname and accumulate into slice
```
I think this comment is no longer accurate :)

pmorie commented on plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go:
```diff
@@ -103,11 +104,13 @@ func (d *defaultServicePlan) Admit(a admission.Attributes) error {
+	// loop over all service plans for the specified classname and accumulate into slice
+	plans, err := d.getClusterServicePlansByClusterServiceClassName(sc.Name)
+	if err != nil {
+		msg := fmt.Sprintf("Error listing plans for service class %q - retry and specify desired ClusterServicePlan", instance.Spec.ExternalClusterServiceClassName)
```
Can you please disambiguate the k8s and external names of the ClusterServiceClass?

jboyd01 commented:
@pmorie good catches.  I've removed the lister, the WaitForReady check,  bogus comment and updated the error with both k8s and internal class name addressing all of your review comments.  Thanks for the review.

vaikas-google commented on plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go:
```diff
@@ -85,7 +78,7 @@ func (d *defaultServicePlan) Admit(a admission.Attributes) error {
 	}
```
This should also check if instance.Spec.ClusterServicePlanName != ""
so: 
if  instance.Spec.ExternalClusterServicePlanName != "" || instance.Spec.ClusterServicePlanName != "" {

See PR:
https://github.com/kubernetes-incubator/service-catalog/pull/1377

Perhaps this got lost in rebases?

jboyd01 replied:
not lost, this PR just needed to be rebased.  Doing it now.

vaikas-google replied:
Odd, I was expecting the little box on the bottom to say that it needed to be rebased but it didn't. Thanks!!

vaikas-google commented:
LGTM with one caveat, surprised the tests passed however?
